File: 238_product_of_array_except_self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
  result = Array.new(nums.length, 1)
  lower_prod = 1
  upper_prod = 1
  
  nums.each_with_index do |num, idx|
    result[idx] *= lower_prod
    result[-(idx+1)] *= upper_prod

    lower_prod *= num
    upper_prod *= nums[-(idx+1)]
  end

  result
end

File: 238_product_of_array_except_self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  let results = [];
  for (let i = 0; i < nums.length; ++i) {
    results.push(1);
  }
  let upperProd = 1;
  let lowerProd = 1;

  for (let i = 0; i < nums.length; ++i) {
    results[i] *= lowerProd;
    results[nums.length - 1 - i] *= upperProd;

    lowerProd *= nums[i];
    upperProd *= nums[nums.length - 1 - i];
  }

  return results;
};